### PR TITLE
Process API

### PR DESCRIPTION
**Pull request by trustmaster**
This is implementation of the process API and features discussed in #205 and #373.

Simple component example can be found in [spec/components/MergeObjects.coffee](https://github.com/trustmaster/noflo/blob/process_api/spec/components/MergeObjects.coffee#L11)
More examples and advanced cases are in the [Component spec](https://github.com/trustmaster/noflo/blob/process_api/spec/Component.coffee#L311)

This is still work in progress and subject to discussion. Here is the tracking checklist:

- [x] Process API
- [x] Controls
- [x] Non-triggering ports
- [x] Scope-based packet isolation
- [x] `ordered` option to maintain the correct sequence between activation and output
- [x] Error handling API

Questions to be discussed:

1. Should components be `ordered: true` or `ordered: false` by default. Maintaining order comes for a price of extra buffering of output before sending it.
2. What's our default policy for buffer bounds? Should buffer bounds be set on graph level, per component or per port?
3. Should we stick with `c.error()` and `error` outport for error handling as before? Or do we know a better way?

### CODE REVIEW

File: src/lib/Component.coffee [hunk 1]
```diff
@@ -56,4 +68,99 @@ class Component extends EventEmitter
...
+      @nodeInstance.outPorts[port].sendIP ip
+
+  send: (outputMap) ->
+    if @nodeInstance.ordered and
+    not ('__resolved' of @result)
+      @activate()
+    for port, packet of outputMap
+      @sendIP port, packet
+
+  sendDone: (outputMap) ->
```
Comment (aretecode) [inline src/lib/Component.coffee]: This will help save some lines!

**Comment by bergie** [comment]
`c.error` is easy to use, but the problem is that it carries a risk of not knowing the right context for the error. This has been especially problematic with MsgFlo where knowing the message scope for ACK/NACK purposes is critical. So, we would need to ensure `c.error` is aware of both brackets and scope.

**Comment by jonnor** [comment]
Most of the time a component should either provide output, or send error. Therefore I propose a node.js style `send(err, outputMap)`.

**Comment by bergie** [comment]
@jonnor kinda makes sense, and sounds like would also make it easier to wrap existing Node.js libraries into NoFlo

**Comment by jonnor** [comment]
I don't think `ordered: true` makes sense as a default. In fact I'm not sure it makes sense to provide as a component-level feature at all.

In classical FBP, there is a single thread executing inside a component, and it can only do one thing at a time. Ordering here comes natural (and it does have some benefits). If one wants concurrency, one instantiate several components, each handling a request at a time.

However, in NoFlo each component executes a (currently boundless) number of input/activations concurrently. Having the component try to enforce order will just make

1. The latency of results on subsequent inputs be increased to the maximum of the inputs before it. This is particularly damaging in timeout/error situations, say one request takes 100x time the other ones -> everything else gets delayed that amount the same.
2. Related, the memory needed will pile up uneccesarily

Ordering like this is in most node.js and browser applications not-necessary. And if it is needed in a particular application, I think it can, and should be handled there - using by wiring in a component.

**Comment by bergie** [comment]
@jonnor agreed, ordering should not be necessary if we still maintain bracket and scope correctly, as you can always put things back together downstream as needed.

**Comment by trustmaster** [comment]
Thanks for comments, that makes sense and aligns with my own opinion.

I have further questions about error handling:

1. What does it mean for `error` to be aware of brackets? Does it mean automatic forwarding of all brackets received prior to activation? Wouldn't it clutter error handling?
2. How about components like validators which can emit a list of errors rather than a single error object?

The the original question on buffer bounds is still open.

**Comment by automata** [comment]
I have some questions after reading the proposal, example and tests:

- First of all, I really like IP envelopes. Having groups (and scopes) in IP class is something I always wanted in Caliper: we have this case where we use groups to identify measurements for a given URL (using URL as group id) and groups keeping growing between components make it easy to mess group ordering. I can imagine that making it sure an IP belongs to a group will make it possible to decide right on the component if we should process the IP of a expected URL, for example. We can probably do that with `begin`/`endGroups` but I think this way is much more explicit.
- I didn't get what `inPorts.foo.control: true` does. Does it flag that the component will send output (or call `process`) only when receiving IPs on `foo` (control) inport?
- What's the difference between `output.sendDone` and `output.send`?
- Are `triggering: false` ports like WirePattern's params? BTW it would be interesting to have a guideline doc describing how to migrate from WirePattern to Process, like some patterns like this with examples. I'd like to contribute to it.
- What's the difference between `triggering: false` and `required: false`?
- In which situation `return unless input.has "foo", "bar"` will be `true`? Is it only a safe check for inports or does it means `process` will continue only when `foo` and `bar` inports are triggered?

Sorry for the number of questions, I don't know if this is the right place to ask them, please point me to a better place if it exists :-)

**Comment by trustmaster** [comment]
@automata all these questions are legit as I haven't bothered writing the docs yet as we haven't discussed this API before.

- `control` ports are input ports which remember last received value. `input.get()` returns a value without moving it off the buffer. So they are the replacement for the old WirePattern `param`
- `triggering: false` is an orthogonal feature which was a part of WirePattern `param` before. Non-triggering ports don't call the `process` handler. They can be control or non-control.
- `return unless input.has "foo", "bar"` makes sure that both `foo` and `bar` inputs have data to be received. This is the way we describe firing patterns in NoFlo 1.0. Here is a more complex example:

```coffeescript
process (input, output) ->
  return unless input.has 'arr' or input.has 'foo', 'bar'
  if input.has 'arr'
    # Direct buffer access
    found = false
    for item, index in input.ports.arr.buffer
      # Find something among all buffered packets
      if item.something is 'foo'
        # Do something
        found = true
      return unless found
  else
    [foo, bar] = input.get 'foo', 'bar
    # ...
```

 - `output.sendDone()` is used to send activation results and deactivate the process with a single line and is recommended in most cases. `output.send()` is used to send intermediate results during the activation and requires calling `done()` to deactivate

**Comment by automata** [comment]
@trustmaster thank you so much for the answers! Going throught them in order:

- so when do `control` inports will get their values removed from buffer? When component disconnects?
- got it, so if an inport is `control: true` and `triggering: false` it will act just like a WirePattern `param` inport, right? Because it will remember last state and will not trigger at same same
- this is really great, I can see it as a more granular way to decide based on inports presence and data received

**Comment by jonnor** [comment]
Buffer bounds can go into a separate PR.

**Comment by trustmaster** [comment]
@automata 

- when `control` receives a new value it moves the previous one off the buffer. So it has a buffer limited to 1 item
- exactly! `control: true + triggering: false == WirePattern.param`

**Comment by trustmaster** [comment]
Error handling is now on board. Proposed ways to send errors include:

```cofeescript
process: (input, output) ->
  output.sendDone new Error 'Should be an error instance'
```

```coffeescript
process: (input, output) ->
  errors = []
  errors.push new Error 'Multiple errors per activation one'
  errors.push new Error 'Multiple errors per activation two'
  output.sendDone errors # Produces a substream
```

Same way of passing errors is supported by `output.send()` and `done()` functions.

This PR is now fine to be merged.

**Pull request merged**